### ISSUE TITLE
Consider defining -last-item and -first-item as generalised variables

### ISSUE THREAD

**Issue by Wilfred** [magnars/dash.el#200]
I'd like to be able to write:

```lisp
(setf (-last-item some-list) new-value)
```

Rather than:

```lisp
(setf (car (last some-list)) new-value)
```

Would you be open to a patch that did this?

**Comment by Fuco1** [comment]
This gives me an idea... maybe we could add this interface also for "predicates" like `-last` or `-first` and similar commands... that would enable us to "find the element and modify it" in a universal way (basically simulating functions like `--replace-first` and similar).

Would such an API make sense?

**Comment by Wilfred** [comment]
I'm not sure what you're describing. Do you mean supporting something like this?

```lisp
(setf (-first #'evenp some-list) 0)
```

**Comment by Fuco1** [comment]
Exactly.  That's equivalent to a hypothetical api `(-replace-first #'evenp some-list 0)` but with the advantage we don't have to add a version for each dash function which does something like this.

The disadvantage is that AFAIU `setf` is side-effect and dash is trying to be pure.  Can you make a pure version of `setf`?

**Comment by daantjie** [comment]
`cl-letf` is pure:

```el
(setq lst '(1 2 3))
(cl-letf (((-first-item lst) 0))
  (copy-list lst)) ;; => (0 2 3)
```

Without the `copy-list`, you'd get `(1 2 3)` back. You could put this into a macro:

```el
(defmacro -copyf (lst place value)
  `(letf ((,place ,value))
     (copy-list lst)))

(-copyf lst (-first-item lst) 0) ;; => (0 2 3)
```

Of course, this is far from perfect, because you have to specify `lst` as well as `place`, meaning that the following won't work:
```el
(-copyf '(1 2 3) (-first-item '(1 2 3)) 0) ;; => (1 2 3)
```

It's a start, anyhow.

**Comment by daantjie** [comment]
You can make a slight improvement, if you let `--copyf` be anaphoric.

```el
(defmacro --copyf (lst place value)
  `(letf* ((it ,lst)
           (,place ,value))
     (copy-list it)))

(--copyf '(1 2 3) (-first-item it) 0) ;; => (0 2 3)
```

I've looked through some of the `gv` code, and I have a feeling that to make something like `-copyf` that doesn't require (effectively) specifying the list twice will be fairly involved (but I'll give it a good shot).

**Comment by Fuco1** [comment]
This was partially implemented in #205
